src/main/kotlin/org/jetbrains/research/testspark/tools/Pipeline.kt

```diff
@@ -69,17 +90,34 @@ class Pipeline(
                     if (projectBuilder.runBuild(ijIndicator)) {
                         if (processStopped(project, ijIndicator)) return
 
-                        processManager.runTestGenerator(
+                        result = processManager.runTestGenerator(
                             ijIndicator,
                             codeType,
                             packageName,
+                            projectContext,
+                            generatedTestsData,
                         )
                     }
 
                     if (processStopped(project, ijIndicator)) return
 
                     ijIndicator.stop()
                 }
+
+                override fun onFinished() {
+                    super.onFinished()
+                    project.service<RunnerService>().clear()
+                    result?.let {
+                        project.service<ReportLockingService>().receiveReport(it)
+                    }
+                }
             })
     }
+
+    fun clear(project: Project) { // should be removed totally!
+        project.service<TestCaseDisplayService>().clear()
+        project.service<ErrorService>().clear()
+        project.service<CoverageVisualisationService>().clear()
+        project.service<TestsExecutionResultService>().clear()
+    }
```

Comment:
Does the removal of this function require rewriting these 4 services? i.e. how much effort/time needed to remove it?

Reply:
We need to clear the data in these services so that we do not show the results from the previous test generation session. Removing this function means that we need to solve this issue in another way. My suggestion for the next refactoring steps is to convert these classes from services to normal classes which are initiated the beginning of each test generation session. In this solution, we don't need to clear any data as in any test generation process, we will have a new instance dedicated to that session.

If we want to implement this suggestion, I would say it needs relatively considerable effort/time. But, maybe not as much as the amount of time we put for removing previous services.